modules/hitl2/src/backend/api.ts

```diff
@@ -0,0 +1,289 @@
+
+      const online = await repository.setAgentOnline(req.params.botId, agentId, true)
+      const payload = {
+        online: online
+      }
+
+      realtime.sendPayload({
+        resource: 'agent',
+        type: 'update',
+        id: agentId,
+        payload: payload
+      })
+
+      res.json(payload)
+    })
+  )
+
+  router.post(
+    '/agents/me/offline',
+    hitlMiddleware(async (req: RequestWithUser, res: Response) => {
+      const { email, strategy } = req.tokenUser!
+      const agentId = makeAgentId(strategy, email)
+
+      const online = await repository.setAgentOnline(req.params.botId, agentId, false)
+      const payload = {
+        online: online
+      }
+
+      realtime.sendPayload({
+        resource: 'agent',
+        type: 'update',
+        id: agentId,
+        payload: payload
+      })
+
+      res.json(payload)
+    })
+  )
+
+  router.get(
+    '/escalations',
+    hitlMiddleware(async (req: Request, res: Response) => {
+      const escalations = await repository.getEscalationsWithComments(
+        req.params.botId,
+        _.pick(req.query, ['limit', 'orderByColumn', 'orderByDirection']) as CollectionConditions
+      )
+      res.json(escalations)
+    })
+  )
+
+  router.post(
+    '/escalations',
+    hitlMiddleware(async (req: Request, res: Response) => {
+      const payload = {
+        ..._.pick(req.body, ['userId', 'userThreadId']),
+        status: 'pending' as 'pending'
+      }
+
+      Joi.attempt(payload, CreateEscalationSchema)
+
+      // Prevent creating a new escalation if one is currently pending or assigned
+      let escalation
+      escalation = await repository
+        .escalationsQuery(builder => {
+          return builder
+            .where('botId', req.params.botId)
+            .andWhere('userId', payload.userId)
+            .andWhere('userThreadId', payload.userThreadId)
+            .whereNot('status', 'resolved')
+            .orderBy('createdAt')
+            .limit(1)
+        })
+        .then(data => _.head(data))
+
+      if (escalation) {
+        res.sendStatus(200)
+      } else {
+        escalation = await repository.createEscalation(req.params.botId, payload).then(escalation => {
+          state.cacheEscalation(req.params.botId, escalation.userThreadId, escalation)
+          return escalation
+        })
+
+        realtime.sendPayload({
+          resource: 'escalation',
+          type: 'create',
+          id: escalation.id,
+          payload: escalation
+        })
+
+        res.status(201)
+        res.json(escalation)
+      }
+    })
+  )
+
+  router.post(
+    '/escalations/:id/assign',
+    hitlMiddleware(async (req: RequestWithUser, res: Response) => {
+      const { email, strategy } = req.tokenUser!
+
+      const agentId = makeAgentId(strategy, email)
+      const online = await repository.getAgentOnline(req.params.botId, agentId)
+
+      Joi.attempt({ online: online }, AgentOnlineValidation)
```

Comment:
Please use a middleware to check if the agent is online, instead of copy-pasting  the same 4 lines when he must be online.Ex:
```js
const agentOnlineMw = fn => {
    return async (req: BPRequest, res: Response, next) => {
      const { email, strategy } = req.tokenUser!
      const agentId = makeAgentId(strategy, email)
      const online = await repository.getAgentOnline(req.params.botId, agentId)

      if (!online) {
        return next('Must be online')
      }

      next()
    }
  }

router.post(
    '/escalations/:id/assign',
   agentOnlineMw,
    hitlMiddleware(async (req: RequestWithUser, res: Response) => {
```